### DISCUSSION TITLE
Outbound Shipment: Add Item

### DISCUSSION THREAD

**Discussion by mark-prins** [openmsupply/openmsupply-client#483] [Show and tell]
With the outbound shipments, when adding an item, it was unclear whether the quantity requested was number of packs or number of units. To try and address this we've changed the automatic allocation of packs slightly, so the process flow is like this:
 
1. you can choose whether to use packs or units
2. then enter a quantity
3. choose the pack size, if you are using packs 
4. click `Allocate` to automatically select batches using FEFO logic, allocating only to batches which have the pack size requested

The window now looks like this:
<img width="947" alt="Screen Shot 2021-11-24 at 10 50 44 AM" src="https://user-images.githubusercontent.com/9192912/143135554-818d183e-0315-454a-8184-f470577fe7bf.png">

I think that the extra click (the Allocate button) is worth the clarity this gives. what about you? Your thoughts?

**Comment by mark-prins** [comment]
Note: when checking on the demo server, there is stock for Amoxicillin and Vancomycin ( see the [stock page](https://demo-open.msupply.org/inventory/stock) for a list )

Note: The API currently forces you to match the case when searching, so looking for `amox` won't find `Amoxicillin`

**Comment by richardmoizeau** [comment]
Following review with @craigdrown, let me start by saying that the overall design and behaviour/logic of the modal is great

Here are our suggestions/comments: 

## headers

![Screen Shot 2021-12-24 at 2 10 21 PM](https://user-images.githubusercontent.com/74992958/147303795-2ef650f7-c6ca-4ae7-a2ba-6468fdfc6676.png)

1. Remove the [Code] and add it to the left of the item name in the [Item] dropdown. 
2. Remove [Available] and [Unit] and replace it by a statement: `Available Quantity: 125 tabs`
3. Remove total [Allocated] field (see next section for the why)

## Allocate to packs?

See: https://docs.google.com/presentation/d/13DXRy6zmp2SLIY5cc_YqGIxIm1Qkh1PrIiaMFULnhzQ/edit#slide=id.p

### Slide 1: 
- Remove the [Allocate to packs?] switch. By default, a user will enter a quantity/number of units: `Issue [Quantity] [Unit] in pack size of [Pack Size]`
- Default value in pack size = `Any`. This means that the quantity of units will be allocated using the FEFO rules regardless of the pack size. 

![Screen Shot 2021-12-24 at 2 42 58 PM](https://user-images.githubusercontent.com/74992958/147305550-cac58ecd-04c8-4d98-aaef-de311f752126.png)

### Slide 2: 
- a user can edit the "automatic allocation" at the batch level in the available batch(es) table.
- value in the issue [Quantity] is automatically updated based on user inputs
- you can reset allocation by entering 0 in the issue [Quantity] field and clicking on Allocate

![Screen Shot 2021-12-24 at 2 43 39 PM](https://user-images.githubusercontent.com/74992958/147305589-75b8f824-90ff-4747-b5f4-41e78c3c2f5b.png)

### Slide 3: 
- now let's imagine that the user wants to issue a number of packs (pack size 25). User clicks on the [Pack Size] dropdown and select the value `25`. 
- value in the issue [Quantity] and at batch levels are reset to `0` (this should happen every-time user changes the value of the [Pack Size] dropdown)
- the sentence changed from `Issue [Quantity] [Unit] in pack size of [Pack Size]` to `Issue [Quantity] packs of [Pack Size] [Units]`
- in the batch(es) table, we could de-activate batch(es) with another pack size than 25
- user enters the quantity of packs and click on [Allocate] and system allocates the quantity to batch(es) with pack size of 25, using the FEFO rule
- user can also edit the allocation and this will also automatically update the issue [Quantity] value

![Screen Shot 2021-12-24 at 2 44 20 PM](https://user-images.githubusercontent.com/74992958/147305632-b7e8fb19-c33c-44e1-9f01-bf22fbe9879e.png)

## Batch(es) Table

- You can see in the slides a suggested order for the fields of the table: batch number > Expiry Date > Number of Packs > Pack Size > Total Units Qty > Available > In Store > Location > Unit Price > On Hold 
- In the slides, I also tried to separate the table in 2 (using a different header color) between quantity data and extra info about the batches (location, unit price) 
- I also added a total line for [Total Unit Qty] but not too sure it is useful

### Questions: 
- what happens to Placeholder quantity when you allocate a number of packs ? a number of units ? Should placeholder always be in number of units ? If you select a pack size of 25, should the placeholder be a quantity of packs of 25 ? 

Thank you in advance for your feedback & suggestions. 

Richard

**Reply by josh-griffin** [reply]
@richardmoizeau 

Thanks!! This is super helpful thanks a lot for spending the time to think through this so deeply.

I'm still thinking it through but some further initial questions:

1. Further with the placeholder, regardless of the quantity value (i.e. your question about the pack size), does the 'allocate' button reduce the placeholder value?
2. It seems that if you use the auto allocate feature, then a user is *only* able to issue to pack sizes of X. I.e. if I allocate 10 packs of pack size 2, then any lines which are not of pack size 2 are disabled. Just checking that this (is correct) and that the trade-off has been considered and accepted.
3. The 'allocate' button seems redundant now if the top level 'allocate' quantity is the sum of issued quantities in the below batch - we can just update the batch rows on the fly without the button

[Side note: point two was the motivation for creating the 'allocate to' button which resets to zero afterward, as it's more of an 'action' to allocate quantites, rather than what you have here which is a single input trying to 'sync up' multiple other inputs]

**Comment by mark-prins** [comment]
Thanks @richardmoizeau ( & @craigdrown )! Great to have the feedback, and to have it so carefully written up!

I suspect the meeting was a little ad-hoc, so this might not be possible.. but wondered if it be possible to include Josh or I next time you have a catch up like this? It would reduce the need to have a long written discussion.

Agree with Josh - the approach you have removes the motivation for the `Allocate` button and if we move to an auto-updated summary line, then there is no need for the button. Is the trade off ok? You could only have per pack allocations in a single pack size? ( I think this is ok, btw )

Some thoughts on reading the above:
- the value we have for `unit` is singular and so it won't work very well in a sentence, as you have it: 'Issue `75` Tab/Vial/Bottle/Capsule' or 'Available quantity: `25` Tab'. Appending 's' will only work in English; will we have to store a plural form in the `[unit]` table now?
- quantity values are auto update one way (from batch table to issue field), but not automatic the other way ( click the `allocate` button ) the inconsistent behaviour is another reason to remove the button :)
- The process of 'allocating' a value of `0` in order to set a requested quantity of zero is odd to me; this fits more with a view of the 'issue' field being a controlling value of the total requested, and so I think you are mixing behaviour here. Again, removing 'allocate' will help
- As mentioned before, and I think this has purely come through from the design process: on hold shouldn't really be a checkbox as you cannot click it. what if we simply had a tick in that column?
- The total row at the bottom - this shows the total units, a value which is already shown when using 'Any' as the pack size. I prefer having text, similar to `Available quantity x units` which shows in what I think of as the summary area ( issue quantity and pack size selection ) and only shows when a specific pack size is selected

And then placeholders.. which is only really on this screen because it was in the designs. I have questions/problems with the placeholder
- can we remove placeholders? Are they required for outbound shipments? 
- can we handle them differently?
- the quantity for the placeholder is in the packs column; when the pack size is 'Any' we can't exactly use 'number of packs'. It was in that column as it looks nicer, but I'm worried that it is confusing
- placeholder quantity was in units, as we were allowing mixed pack sizes ( and still are, if you choose 'Any' ). When choosing 'Any' which pack size should the placeholder be?
- In current mSupply, the requested pack size is part of the placeholder row. should we do this too?

I think placeholders will need a separate discussion. 😬

**Comment by richardmoizeau** [comment]
Minutes of the 29/12 meeting: https://docs.google.com/document/d/1yoFZVfa0309AV660XHtDJhHAfsZqJaPgp6wdQ7DbZTU/edit# with @joshxg, @mark-prins, @craigdrown and myself. 

Also below mockups of the 'new' modal:
- when issuing quantity of units: 
![Screen Shot 2021-12-29 at 4 08 41 PM](https://user-images.githubusercontent.com/74992958/147623733-aefac14a-f1fd-4c64-9944-5805236d58ca.png)
- when issuing quantity of packs: 
![Screen Shot 2021-12-29 at 4 09 07 PM](https://user-images.githubusercontent.com/74992958/147623748-e60c1a12-4139-43c0-bab1-e69a2fea427f.png)
